auth/auth.go

```diff
@@ -223,13 +224,47 @@ func (auth *Authenticator) getPrincipal(docID string, factory func() Principal)
 	return princ, nil
 }
 
+func (auth *Authenticator) channelCount(princ Principal, user User, roles []Role, scope, collection string) int {
+	cumulativeChannels := 0
+
+	channels := princ.CollectionChannels(scope, collection).Copy()
+	for _, role := range roles {
+		roleSince := user.RoleNames()[role.Name()]
+		channels.AddAtSequence(role.CollectionChannels(scope, collection), roleSince.Sequence)
+	}
+	// We want to keep track of cumulative number of channels for limits across all collections
+	cumulativeChannels += len(channels)
+	return cumulativeChannels
+}
+
+func (auth *Authenticator) getUserRoles(user User) ([]Role, error) {
+	roles := make([]Role, 0, len(user.RoleNames()))
+	for name := range user.RoleNames() {
+		role, err := auth.GetRole(name)
+		if err != nil {
+			return nil, err
+		} else if role != nil {
+			roles = append(roles, role)
+		}
+	}
+	return roles, nil
+}
+
 // Rebuild channels computes the full set of channels for all collections defined for the authenticator.
 // For each collection in Authenticator.collections:
 //   - if there is no CollectionAccess on the principal for the collection, rebuilds channels for that collection
 //   - If CollectionAccess on the principal has been invalidated, rebuilds channels for that collection
 func (auth *Authenticator) rebuildChannels(princ Principal) (changed bool, err error) {
 
 	changed = false
+	cumulativeChannels := 0
```

Comment:
declare as uint32 so you don't have to cast it?

Reply:
I'll do this but will have to cast int to uint32 in the channelCount function